packages/core/src/format/format.ts

```diff
@@ -126,6 +126,33 @@ export function formatProblems(
         process.stderr.write('\n');
       }
       break;
+    }
+    case 'checkstyle': {
+      const groupedByFile = groupByFiles(problems);
+      const writeln = (line: string) => process.stdout.write(`${line}\n`);
+
+      writeln('<?xml version="1.0" encoding="UTF-8"?>');
+      writeln('<checkstyle version="4.3">');
+
+      for (const [file, { fileProblems }] of Object.entries(groupedByFile)) {
+        writeln(`<file name="${xmlEscape(path.relative(cwd, file))}">`);
+        for (let i = 0; i < fileProblems.length; i++) {
+          const problem = fileProblems[i];
+          const { line, col } = problem.location[0].start;
+          const severity = problem.severity == 'warn' ? 'warning' : 'error';
+          const message = xmlEscape(problem.message);
+          const source = xmlEscape(problem.ruleId);
+          writeln(
+            `<error line="${line}" column="${col}" severity="${severity}" message="${message}" source="${source}" />`,
+          );
```

Comment:
Could we move this into a separate function `formatCheckstyle` similar to `stylish` to make it consistent? Also, switch to `for of` instead of regular for loop.

Reply:
I somehow felt that, without a function, it's easier to grasp the report file format, given that it's just a few lines.

Reply:
Oh, then I also have to toss away the writeln arrow function.

Reply:
@twz123 I agree it's somehow easier but I from the other perspective consistency also is important.

You should extract writeln to the module scope. It doesn't have any closure deps.